agents_infra/agents_infra/utils/communication.py

```diff
@@ -0,0 +1,32 @@
+from logtools import maybe_log_message
+from token_manager import TokenManager
+
+
+class AgentCommunicator(object):
+    def __init__(self, server_name, post_data_fn, logger):
+        self.server_name = server_name
+        self.post_data = post_data_fn
+        self.logger = logger
+        self.token_manager = TokenManager(server_name, post_data_fn, logger)
+
+    def post_data_with_auth(self, url, payload, **kwargs):
+        token = self.token_manager.get_valid_token()
+        if not token:
+            maybe_log_message(
+                'Token retrieval failed',
+                logger=self.logger,
+                extra={
+                    'agent_name': self.server_name,
+                    'step': 'auth',
+                    'status': 'missing_token'
+                }
+            )
+            raise Exception('No valid token available for authentication')
```

Comment:
Use a more specific exception class rather than the base Exception, e.g., AuthenticationError or a custom error.
Consider making the error message more descriptive, such as including server_name or relevant context.
Make sure this error is handled at a higher level so it doesn't crash the whole application unexpectedly.